### PR TITLE
Add commandline flag to set genesis block script destination

### LINKED ISSUE #54: Regtest needs ability to change the blocksigning scriptpubkey from the command line

### PR DESCRIPTION
**Pull request by apoelstra**
Fixes #54

### CODE REVIEW

**Comment by jtimon** [comment]
Why not call ScriptDestinationFromCommandLine etc directly in CRegTestParams instead of in SelectParamsFromCommandLine() ?
Or maybe CRegTestParams's constructor could take a CScript parameter that defaults to OP_TRUE. 
Certainly ChangeGenesisScriptDestination doesn't need to be a public method, all this can be a just a static function in chainparams.cpp
In fact, this PR is buggy since ChangeGenesisScriptDestination doesn't update hashGenesisBlock nor mapCheckpointsTestnet[0].

**Comment by apoelstra** [comment]
@jtimon For "why did I structure the functions this way" I was trying to copy as best I could what is done for the `-testnet`/`-regtest` parameters. As for giving a parameter to the `CRegTestParams` that's tricky because the regtest paramaters are a static variable, and IIRC in C++ we have no control over what order constructors for static variables are run in.

As for the bugs, you're right, thanks.

**Comment by jtimon** [comment]
re CRegTestParams::CRegTestParams() parameter: I was assuming some of https://github.com/bitcoin/bitcoin/pull/6382 had already been merged, but some of those changes were rejected for alpha so never mind.

Anyway, the solution is still simple: just put all the new code inside CRegTestParams::CRegTestParams(), before the hash of the genesis block is calculated.

See https://github.com/bitcoin/bitcoin/pull/6382 which also builds the genesis block dynamically from a command line parameter for the sizetest ```std::numeric_limits<uint64_t>::max()``` different chains.

File: src/chainparams.cpp [hunk 1]
```diff
@@ -249,7 +249,7 @@ class CRegTestParams : public CTestNetParams {
         txNew.vout[1].scriptPubKey = CScript() << OP_TRUE;
         genesis.vtx[0] = CTransaction(txNew);
 
-        CScript scriptDestination(CScript() << OP_TRUE);
+        CScript scriptDestination(CScript() << OP_2 << ParseHex("023303dedc51b9d227b17c9fb4710f96b844e1ccdc2c776e1b7274bd4e246b6202") << ParseHex("03024c3b4f830854d6d26d6e34d92aff4c703bf57e85cd42abe328d928e01d4286") << ParseHex("03fe4e8c8d99b9dcbb529a87d54f606bae0149a34018325547fa0c2239e038a1c9") << OP_3 << OP_CHECKMULTISIG);
```
Comment (apoelstra) [inline src/chainparams.cpp]: Oops, this should not have made it into the PR.

**Comment by apoelstra** [comment]
Thanks for https://github.com/bitcoin/bitcoin/pull/6382/files @jtimon, there are a lot of good ideas in there. I think I'll copy your factory pattern.

**Comment by jtimon** [comment]
Awesome! Maybe you want to utACK one of the following open little steps in bitcoin/bitcoin?

 Chainparams: Translations: DRY: options and error strings #6235
 Don't check the genesis block #6597

**Comment by apoelstra** [comment]
Rebased using a factory pattern as in @jtimon's code. Need to add some unit tests before this is usable.

Like bitcoin/6382, there is an argument that this is only useful for QA and may not make sense to be merged. But in this case, I think it's reasonable to expect users to play with their own network with their own functionaries, and this makes this much easier.

**Comment by apoelstra** [comment]
Added unit tests that confirms the genesis hash is computed correctly (for testnet and regtest, that the output of RPC `getblockhash 0` matches the result of directly passes `CScript`s into `SetParams`; for mainnet and unittest I assumed the result was correct and added a check for it to catch regressions).

**Comment by apoelstra** [comment]
@jtimon Do you have further thoughts on this PR? I know you have overlapping PRs in Core.

@gmaxwell You commented on https://github.com/bitcoin/bitcoin/pull/6907 that you feel making this a factory is a bad idea. Keeping them static would mean moving the constructor into some sort of `init` function, have the constructor just call this, and have `SelectParams` call that rather than calling the factory. Would you prefer this?

In any case I'd like to move forward with this PR.

File: src/chainparams.h [hunk 2]
```diff
@@ -80,6 +80,14 @@ class CChainParams
     const std::vector<unsigned char>& Base58Prefix(Base58Type type) const { return base58Prefixes[type]; }
     const std::vector<CAddress>& FixedSeeds() const { return vFixedSeeds; }
     virtual const Checkpoints::CCheckpointData& Checkpoints() const = 0;
+    /** Change the script in the genesis hash */
+    void ChangeGenesisScriptDestination(CScript s) { genesis.proof = CProof(s, CScript()); }
```
Comment (jtimon) [inline src/chainparams.h]: Unused method.

File: src/chainparams.cpp [hunk 3]
```diff
@@ -80,7 +80,7 @@ static const Checkpoints::CCheckpointData dataRegtest = {
 
 class CMainParams : public CChainParams {
 public:
-    CMainParams() {
+    CMainParams(CScript scriptDestination) {
```
Comment (jtimon) [inline src/chainparams.cpp]: Is this ever used for MAIN ?

File: src/chainparams.cpp [hunk 4]
```diff
@@ -308,16 +309,14 @@ class CUnitTestParams : public CMainParams, public CModifiableParams {
     virtual void setAllowMinDifficultyBlocks(bool afAllowMinDifficultyBlocks) {  fAllowMinDifficultyBlocks=afAllowMinDifficultyBlocks; }
     virtual void setSkipProofOfWorkCheck(bool afSkipProofOfWorkCheck) { fSkipProofOfWorkCheck = afSkipProofOfWorkCheck; }
 };
-static CUnitTestParams unitTestParams;
 
-
-static CChainParams *pCurrentParams = 0;
+static CChainParams *pCurrentParams;
 
 CModifiableParams *ModifiableParams()
```
Comment (jtimon) [inline src/chainparams.cpp]: Is this function still being used anywhere?

**Comment by jtimon** [comment]
If I understood it correctly, @gmaxwell 's nit was that I was making globalChainParams and globalSwitchingChainParams extern instead of static and it's already solved (but it would be nice if he can clarify whether his nit has been solved or not in https://github.com/bitcoin/bitcoin/pull/6907 ).

As said, I don't like the script being passed to SeelectParams(), the factory, nor the constructors and I would prefer ScriptDestinationFromCommandLine to be called directly within the constructors, but just a preference.
I added a couple of minor nits though.

The overlapping PR/commits in Core are contained in https://github.com/jtimon/bitcoin/tree/alpha-base-0.12 which I will maintain rebased until 0.12 is forked and is what I would l would like @TheBlueMatt to use as a base when rebasing.

File: src/chainparams.cpp (hunk 3 above)
Comment (apoelstra) [inline src/chainparams.cpp]: Is MAIN ever used period? I think this should be there for symmetry's sake in any case.

File: src/chainparams.cpp (hunk 4 above)
Comment (apoelstra) [inline src/chainparams.cpp]: Apparently not. I'll remove it.

**Comment by apoelstra** [comment]
Addressed @jtimon's nits

File: src/chainparams.h (hunk 2 above)
Comment (apoelstra) [inline src/chainparams.h]: Removed

File: src/chainparams.cpp (hunk 3 above)
Comment (jtimon) [inline src/chainparams.cpp]: Yes, in all unit tests, but fair enough.

**Comment by jtimon** [comment]
I'm going to merge.

File: src/chainparams.h [hunk 5]
```diff
@@ -139,13 +145,15 @@ const CChainParams &Params();
 /** Return parameters for the given network. */
 CChainParams &Params(CBaseChainParams::Network network);
```
Comment (jtimon) [inline src/chainparams.h]: Wait a minute...where is this function defined now?

Comment (apoelstra) [inline src/chainparams.h]: chainparams.cpp:315

Comment (jtimon) [inline src/chainparams.h]: No, I mean the parametrized version.

Comment (apoelstra) [inline src/chainparams.h]: Oh, hmm, this is not defined anywhere anymore (nor is it called anywhere, of course, else we would have compilation errors).

Comment (jtimon) [inline src/chainparams.h]: mhmm I would like to eventually remove it from core too.I guess when rebasing the remaining uses can be replaced with the factory i elements. It's just something I expected to happen in bitcoin core first.

Comment (jtimon) [inline src/chainparams.h]: Of course another option is to maintain it in elements until it removed in core.

**Pull request merged**